atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -120,71 +131,110 @@ public synchronized void storeUpperLimit(final long limit) {
         clientPool.runWithRetry(new FunctionCheckedException<Client, Void, RuntimeException>() {
             @Override
             public Void apply(Client client) {
-                cas(client, currentLimit, limit);
+                cas(client, makeColumnWithNewFormat(getId(), currentLimit), currentLimit, limit);
                 return null;
             }
         });
     }
 
-    private void cas(Client client, Long oldVal, long newVal) {
-        final CASResult result;
+    private void cas(Client client, Column oldColumn, Long oldVal, long newVal) {
+        CASResult result = updateTimestampInDb(client, oldColumn, oldVal, newVal);
+        if (result.isSuccess()) {
+            DebugLogger.logger.info("[CAS] Setting cached limit to {}.", newVal);
+            currentLimit = newVal;
+        } else {
+            if (result.getCurrent_values().isEmpty()) {
+                log.error("[CAS] Error trying to set timestamp limit to {}: there is no limit stored in DB.", newVal);
+                throw new MultipleRunningTimestampServiceError("Error trying to set timestamp limit: there is no "
+                        + "limit stored in DB.");
+            }
+            IdAndTimestamp currentIdAndTimestamp = new IdAndTimestamp(result);
+            /*
+             * For the cas to succeed, the existing entry in the DB must be this.id_this.currentLimit. If that is not
+             * the case, we still want to succeed if:
+             *   1. id in the DB equals this.id -- indicates miscommunication about what was written to the DB,
+             *   but is not a case of multiple running timestamps; or
+             *   2. limit in the DB equals the expected limit, but there is no id/id does not match -- this is the case
```

Comment:
It is consistent with previous behaviour. 

We can avoid this by setting a flag the first time we successfully update the bound that then makes case 2 throw a MRTSE -- @jboreiko does that sound more reasonable?